File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        unique_chars_map = defaultdict(int)
        left, right = 0, 0
        res = 0
        while(right < len(s)):
            curr_char = s[right]
            unique_chars_map[curr_char] += 1
            if unique_chars_map[curr_char] > 1:
                while(unique_chars_map[curr_char] > 1):
                    unique_chars_map[s[left]] -= 1
                    left += 1
            res = max(res, (right - left + 1))
            right += 1
        return res

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.cpp
class Solution {
    int calShipmentDays(vector<int>& weights, int curWeight){
        int res = 0;
        int reminder = 0;
        for (const auto& weight: weights){
            if (reminder + weight > curWeight){
                res += 1;
                reminder = weight;
            }
            else{
                reminder += weight;
            }
        }
        return (reminder == 0) ? res : res + 1;
    }
public:
    int shipWithinDays(vector<int>& weights, int days) {
        if (weights.empty() || days <= 0){
            return -1;
        }
        int left = *max_element(weights.begin(), weights.end());
        int right = accumulate(weights.begin(), weights.end(), 0);
        int mid = 0;
        int res = -1;
        while (left <= right){
            mid = left + ((right - left) / 2);
            if (calShipmentDays(weights, mid) <= days){
                res = mid;
                right = mid - 1;
            }
            else{
                left = mid + 1;
            }
        }
        return res;
    }
};